### ISSUE TITLE
Multithreaded "xbps-query -o" output

### ISSUE THREAD

**Issue by chneukirchen** [voidlinux/xbps#147]
xbps-query -o uses multiple cores, but the lines are generated from three printf calls each, which makes them possibly mix up in the output.

This currently breaks xlocate on more than one core, which makes it slow.

There are two easy ways to fix this:

1) Lock stdout during the writes:

```
From 5564b80a8fa2056ee5d4df96eba9c7bb5027f231 Mon Sep 17 00:00:00 2001
From: Christian Neukirchen <chneukirchen@gmail.com>
Date: Sun, 17 Jan 2016 16:16:28 +0100
Subject: [PATCH] xbps-query: lock stdout for -o output to not mix up writes
 from multiple threads.

---
 bin/xbps-query/ownedby.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/bin/xbps-query/ownedby.c b/bin/xbps-query/ownedby.c
index 40e0e69..a28a32e 100644
--- a/bin/xbps-query/ownedby.c
+++ b/bin/xbps-query/ownedby.c
@@ -76,17 +76,21 @@ match_files_by_pattern(xbps_dictionary_t pkg_filesd,
 		xbps_dictionary_get_cstring_nocopy(obj, "target", &tgt);
 		if (ffd->rematch) {
 			if (regexec(&ffd->regex, filestr, 0, 0, 0) == 0) {
+				flockfile(stdout);
 				printf("%s: %s ", pkgver, filestr);
 				if (tgt)
 					printf("-> %s ", tgt);
 				printf("(%s)\n", typestr);
+				funlockfile(stdout);
 			}
 		} else {
 			if ((fnmatch(ffd->pat, filestr, FNM_PERIOD)) == 0) {
+				flockfile(stdout);
 				printf("%s: %s ", pkgver, filestr);
 				if (tgt)
 					printf("-> %s ", tgt);
 				printf("(%s)\n", typestr);
+				funlockfile(stdout);
 			}
 		}
 	}
-- 
2.7.0
```

2) Refactor code to use a single printf (which locks internally)

```
From 69db088dddbb1fbc7b9f501712a8eb19ca4d8a9f Mon Sep 17 00:00:00 2001
From: Christian Neukirchen <chneukirchen@gmail.com>
Date: Sun, 17 Jan 2016 16:20:34 +0100
Subject: [PATCH] xbps-query: generate -o output from a single printf to not
 mix up threads.

---
 bin/xbps-query/ownedby.c | 18 ++++++++++--------
 1 file changed, 10 insertions(+), 8 deletions(-)

diff --git a/bin/xbps-query/ownedby.c b/bin/xbps-query/ownedby.c
index 40e0e69..fc8684d 100644
--- a/bin/xbps-query/ownedby.c
+++ b/bin/xbps-query/ownedby.c
@@ -76,17 +76,19 @@ match_files_by_pattern(xbps_dictionary_t pkg_filesd,
 		xbps_dictionary_get_cstring_nocopy(obj, "target", &tgt);
 		if (ffd->rematch) {
 			if (regexec(&ffd->regex, filestr, 0, 0, 0) == 0) {
-				printf("%s: %s ", pkgver, filestr);
-				if (tgt)
-					printf("-> %s ", tgt);
-				printf("(%s)\n", typestr);
+				printf("%s: %s%s%s (%s)\n",
+					pkgver, filestr,
+					tgt ? " -> " : "",
+					tgt ? tgt : "",
+					typestr);
 			}
 		} else {
 			if ((fnmatch(ffd->pat, filestr, FNM_PERIOD)) == 0) {
-				printf("%s: %s ", pkgver, filestr);
-				if (tgt)
-					printf("-> %s ", tgt);
-				printf("(%s)\n", typestr);
+				printf("%s: %s%s%s (%s)\n",
+					pkgver, filestr,
+					tgt ? " -> " : "",
+					tgt ? tgt : "",
+					typestr);
 			}
 		}
 	}
-- 
2.7.0

```

**Comment by chneukirchen** [comment]
A quick look at the code showed that this is only an issue for `-o` apparently.

**Comment by xtraeme** [comment]
I prefer your second option, a single printf() seems to be simpler.

Thanks!

**Comment by xtraeme** [comment]
@chneukirchen please commit this yourself, thanks.

**Comment by xtraeme** [comment]
Do not forget to update NEWS.

**Comment by chneukirchen** [comment]
Fixed in 034d359.

**Comment by chneukirchen** [comment]
Fixed in 034d359.
